ci: add arm build for core unix and legacy emulators

vdovhanych opened the pull request:
This PR adds a build for arm emulators needed for trezor-user-env running on arm version of nix. See why [here](https://github.com/trezor/trezor-user-env/pull/114).
This is now done with our m1 runner as docker executor.  I used my nixos image as nixos/nix is completely broken since they switched from alpine to native nixos one. See [here](https://github.com/NixOS/nix/issues/5777#issuecomment-996242619). Once the upstream image is fixed, we can change to that.

This pr also adds native fixes manual macOS build using our runner and built on x86_64 via rosseta2.

matejcik reviewed (approved):
utACK

prusnak commented on ci/build.yml:
```diff
@@ -153,6 +153,22 @@ core unix frozen debug build:
     expire_in: 1 week
 
+core unix frozen debug build arm:
+  image: vdovhanych/nixos
```
this should work now (new arm64 images have been pushed to docker hub recently:

```suggestion
  image: nixos/nix:latest-arm64
```

vdovhanych replied:
This is the one that is broken (nix 2.5.0) that was pushed to docker hub 3 days ago. The latest image was reverted to old nix on alpine (2.3.16) which doesn't have an arm version.

prusnak replied:
OK

prusnak commented on ci/build.yml:
```diff
@@ -301,6 +317,24 @@ legacy emu regular debug build:
     expire_in: 1 week
 
+legacy emu regular debug build arm:
+  image: vdovhanych/nixos
```
same here:

```suggestion
  image: nixos/nix:latest-arm64
```

vdovhanych commented:
The new nixos/nix image was just pushed, I will change it to that and we can merge after that.

prusnak commented:
> The new nixos/nix image was just pushed

`latest` alias still points to `2.5.0` and not to the fixed `2.5.1` - I am asking on Matrix whether this is intentional

vdovhanych commented:
So the new image of `nixos/nix` is unable to build core emulator as the image is missing `/var/tmp` folder 👍 the same goes for tenv as its also unable to be built without creating that folder in `Dockerfile`.

prusnak commented:
> we can just make our base ci image multi-arch one and use that later on.

What would we need to do to make our CI image multi-arch? The Dockerfile is already prepared for multiarch (you can change the arch via the `ALPINE_ARCH` arg).

vdovhanych commented:
> > we can just make our base ci image multi-arch one and use that later on.
> 
> What would we need to do to make our CI image multi-arch? The Dockerfile is already prepared for multiarch (you can change the arch via the `ALPINE_ARCH` arg).

I'm unsure if it will be as simple as adding buildx and setting two targets to build. We would need to modify the `Dockerile` a bit.  The image [`vdovhanych/nixos`](https://github.com/vdovhanych/nixos-docker-arm64/blob/master/Dockerfile) is pretty similar to our base ci image, so there shouldn't be an issue with that.

prusnak commented:
Ok. Let's merge this with `vdovhanych/nixos` as base image for arm64 emulator builds and discuss how to optimize later. Thanks

tsusanka reviewed (commented):
LGTM for now, but as discussed we need to discuss our approach to arm a bit more "conceptually". But let's keep that for Q1 :).

Btw is there a reason to run the jobs for every branch? I would run it for schedules and 
 `/^release\//` /  `/^secfix\//` branches only?

Agree that we don't need to run it on every branch. Should be fine to have it run only on master. as we want to have latest master emulators in tenv right ?

vdovhanych commented:
> LGTM for now, but as discussed we need to discuss our approach to arm a bit more "conceptually". But let's keep that for Q1 :).
> 
> Btw is there a reason to run the jobs for every branch? I would run it for schedules and `/^release\//` / `/^secfix\//` branches only?

Agree that we don't need to run it on every branch. Should be fine to have it run only on master. as we want to have latest master emulators in tenv right ?

tsusanka commented:
Yes, master and also releases/secfix which we want to be deployed to firmware.corp servers (tenv fetches the firmwares from there so new firmware versions are added automatically).

Scheduled is not needed, although I like that it gives us information that "everything is fine from the CI point of view".

vdovhanych commented:
> Yes, master and also releases/secfix which we want to be deployed to firmware.corp servers (tenv fetches the firmwares from there so new firmware versions are added automatically).
> 
> Scheduled is not needed, although I like that it gives us information that "everything is fine from the CI point of view".

Updated the configuration and also added the jobs for releasing arm emulators to our server.

tsusanka commented on ci/deploy.yml:
```diff
@@ -195,6 +195,27 @@ release core unix debug deploy:
+  except:
+    - branches  # run for tags only
+  tags:
+    - deploy
```
Sorry for delaying the PR but is there a reason not to make it part of `release core unix debug deploy`? We can _need_ both of the builds and deploy at the same time. Yes, it will break if one of the builds fails, but I think that is acceptable. We should try to make the jobs compact.

vdovhanych replied:
Both artifacts have the same folder structure and naming, not sure if it won't get overwritten. My first idea was to add it there, but I think it won't work like that.

tsusanka replied:
Oh, I see. Hmm, let's rename it to `*-arm` after the build here? 
https://github.com/trezor/trezor-firmware/pull/2025/files#diff-fad8dced0db200747490daa0aa2fb66c1d30378b9cf5ba032ae96e2e5d8825c1R168

vdovhanych replied:
I tried to do it like that but I'm not sure if its ok that it will overwrite other files but I think that this way we should have both emualtors and we should be able to deploy it with one job.

tsusanka replied:
What other files it would overwrite?

vdovhanych replied:
For the core, we have all of the contents of `core/build/unix` but that is not needed for emulator or tenv. Should be fine 👍

tsusanka commented on ci/deploy.yml:
```diff
@@ -182,33 +182,16 @@ release core unix debug deploy:
     - DEST="$DEPLOY_PATH/trezor-emu-core-v$VERSION"
-    - echo "Deploying to $DEST"
+    - DEST_ARM="$DEPLOY_PATH/trezor-emu-core-arm-v$VERSION"
+    - echo "Deploying to $DEST and "$DEST_ARM"
```
echo "Deploying to $DEST and $DEST_ARM"

(extra ")

tsusanka replied:
Feel free to force-push then!

vdovhanych replied:
fixed in 6f7b404d0
